src/System.Private.CoreLib/src/System/Threading/Thread.CoreRT.cs

```diff
@@ -507,5 +534,65 @@ public static int GetCurrentProcessorId()
                 return RefreshCurrentProcessorId();
             return (currentProcessorIdCache >> ProcessorIdCacheShift);
         }
+
+        internal static void IncrementRunningForeground() 
+        {
+            Interlocked.Increment(ref s_foregroundRunningCount);
+        }
+
+        internal static void IncrementRunningForeground(Thread thread) 
+        {
+            if (!thread.IsBackground) 
+            {
+                IncrementRunningForeground();
+            }
+        }
+
+        internal static void DecrementRunningForeground() 
+        {
+            if (Interlocked.Decrement(ref s_foregroundRunningCount) == 0) 
+            {
+                // Interlocked.Decrement issues full memory barrier 
+                // so most recent write to s_allDone should be visible here
+                s_allDone?.Set();
+            }
+        }
+
+        internal static void DecrementRunningForeground(Thread thread) 
+        {
+            if (!thread.IsBackground) 
+            {
+                DecrementRunningForeground();
+                Debug.Assert(s_foregroundRunningCount >= 1, "s_foregroundRunningCount is less than 1");
+            }
+        }
+
+        internal static void WaitForForegroundThreads() 
+        {         
+            Thread.CurrentThread.IsBackground = true;
+            // last read/write inside `IsBackground` issues full barrier no matter of logic flow
+            // so we can just read `s_foregroundRunningCount`
+            if (s_foregroundRunningCount == 0)
+            {
+                // current thread is the last foreground thread, so let the runtime finish
+                return;
+            }
+            else
+            {
+                Volatile.Write(ref s_allDone, new ManualResetEvent(false));
+                // ManualResetEvent involves kernel transition 
+                // thus other foreground threads could have their job finished meanwhile
```

Comment:
```suggestion
                // Other foreground threads could have their job finished meanwhile
```
This would be a problem even if there was no kernel transition inside `ManualResetEvent`